crypto/hpke/hpke.c

```diff
@@ -292,25 +308,322 @@ static int x25519_auth_decap(const EVP_HPKE_KEY *key,
+#define MLKEM768IPD_INNER_SECRET_KEY_BYTES (384 * 3)
+#define MLKEM1024IPD_INNER_SECRET_KEY_BYTES (384 * 4)
+static void mlkem768_public_from_private(uint8_t *public_key,
+                                         const uint8_t *secret_key) {
+  OPENSSL_memcpy(public_key, secret_key + MLKEM768IPD_INNER_SECRET_KEY_BYTES,
+                 MLKEM768IPD_PUBLIC_KEY_BYTES);
+}
+static void mlkem1024_public_from_private(uint8_t *public_key,
+                                          const uint8_t *secret_key) {
+  OPENSSL_memcpy(public_key, secret_key + MLKEM1024IPD_INNER_SECRET_KEY_BYTES,
+                 MLKEM1024IPD_PUBLIC_KEY_BYTES);
+}
+// PQT secret key is the concatenation of the ML-KEM secret key, EC secret key,
+// and EC public key. So, first extract ML-KEM public key, then extract EC
+// public key.
+static void pqt25519_public_from_private(uint8_t *public_key,
+                                         const uint8_t *secret_key) {
+  mlkem768_public_from_private(public_key, secret_key);
+  OPENSSL_memcpy(
+      public_key + MLKEM768IPD_PUBLIC_KEY_BYTES,
+      secret_key + MLKEM768IPD_SECRET_KEY_BYTES + T25519_SECRET_KEY_BYTES,
+      T25519_PUBLIC_KEY_BYTES);
+}
+static void pqt256_public_from_private(uint8_t *public_key,
+                                       const uint8_t *secret_key) {
+  mlkem768_public_from_private(public_key, secret_key);
+  OPENSSL_memcpy(
+      public_key + MLKEM768IPD_PUBLIC_KEY_BYTES,
+      secret_key + MLKEM768IPD_SECRET_KEY_BYTES + T256_SECRET_KEY_BYTES,
+      T256_PUBLIC_KEY_BYTES);
+}
+static void pqt384_public_from_private(uint8_t *public_key,
+                                       const uint8_t *secret_key) {
+  mlkem1024_public_from_private(public_key, secret_key);
+  OPENSSL_memcpy(
+      public_key + MLKEM1024IPD_PUBLIC_KEY_BYTES,
+      secret_key + MLKEM1024IPD_SECRET_KEY_BYTES + T384_SECRET_KEY_BYTES,
+      T384_PUBLIC_KEY_BYTES);
+}
+
+static int lckem_init_key(EVP_HPKE_KEY *key, const uint8_t *priv_key,
+                          size_t priv_key_len) {
+  const KEM *lckem = KEM_find_kem_by_nid(key->kem->internal_kem_nid);
+  if (lckem == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_UNSUPPORTED_ALGORITHM);
+    return 0;
+  }
+  if (priv_key_len != key->kem->private_key_len) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+    return 0;
+  }
+
+  OPENSSL_memcpy(key->private_key, priv_key, priv_key_len);
+  key->kem->public_from_private(key->public_key, priv_key);
+  return 1;
+}
+
+static int lckem_generate_key(EVP_HPKE_KEY *key) {
+  const KEM *lckem = KEM_find_kem_by_nid(key->kem->internal_kem_nid);
+  if (lckem == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_UNSUPPORTED_ALGORITHM);
+    return 0;
+  }
+  return lckem->method->keygen(key->public_key, key->private_key);
+}
+
+// Derive a shared secret by hashing the KEM shared secret with the KEM
+// ciphertext and KEM public key. This ensures that the resulting KEM is
+// sufficiently binding.
```

Comment:
is there a reference for this?

Reply:
https://datatracker.ietf.org/doc/draft-connolly-cfrg-hpke-mlkem/ does this specifically for ML-KEM and it only hashes the ciphertext, but if you want to extend to all KEMs, then ideally you should hash both the public key and the ciphertext.